internal/controllers/objectsets/objectset_controller.go

```diff
@@ -0,0 +1,236 @@
+		client: c,
+		phaseReconciler: objectsetphases.NewPhaseReconciler(
+			scheme, c, dw, ownerhandling.NewNative(scheme),
+		),
+	}
+	controller.teardownHandler = phasesReconciler
+
+	controller.reconciler = []reconciler{
+		&revisionReconciler{
+			scheme:       scheme,
+			client:       c,
+			newObjectSet: newObjectSet,
+		},
+		phasesReconciler,
+	}
+
+	return controller
+}
+
+func (c *GenericObjectSetController) SetupWithManager(mgr ctrl.Manager) error {
+	objectSet := c.newObjectSet(c.scheme).ClientObject()
+	objectSetPhase := c.newObjectSetPhase(c.scheme).ClientObject()
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(objectSet).
+		Owns(objectSetPhase).
+		Watches(c.dw.Source(), &handler.EnqueueRequestForOwner{
+			OwnerType:    objectSet,
+			IsController: false,
+		}).
+		Complete(c)
+}
+
+func (c *GenericObjectSetController) Reconcile(
+	ctx context.Context, req ctrl.Request,
+) (ctrl.Result, error) {
+	log := c.log.WithValues("ObjectSet", req.String())
+	defer log.Info("reconciled")
+	ctx = logr.NewContext(ctx, log)
+
+	objectSet := c.newObjectSet(c.scheme)
+	if err := c.client.Get(
+		ctx, req.NamespacedName, objectSet.ClientObject()); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if meta.IsStatusConditionTrue(*objectSet.GetConditions(), corev1alpha1.ObjectSetArchived) {
+		// We don't want to touch this object anymore.
+		return ctrl.Result{}, nil
+	}
+
+	if !objectSet.ClientObject().GetDeletionTimestamp().IsZero() ||
+		objectSet.IsArchived() {
+		if err := c.handleDeletionAndArchival(ctx, objectSet); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		objectSet.UpdateStatusPhase()
+		// this controller owns status alone, so we can always update it without optimistic locking.
+		objectSet.ClientObject().SetResourceVersion("")
+		if err := c.client.Status().Patch(ctx, objectSet.ClientObject(), client.Merge); err != nil {
+			return ctrl.Result{}, fmt.Errorf("updating ObjectSet status: %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if err := c.ensureCacheFinalizer(ctx, objectSet); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var (
+		res ctrl.Result
+		err error
+	)
+	for _, r := range c.reconciler {
+		res, err = r.Reconcile(ctx, objectSet)
+		if err != nil || !res.IsZero() {
+			break
+		}
+	}
+	if err != nil {
+		return res, err
+	}
+
+	objectSet.UpdateStatusPhase()
+	// this controller owns status alone, so we can always update it without optimistic locking.
+	objectSet.ClientObject().SetResourceVersion("")
+	if err := c.client.Status().Patch(ctx, objectSet.ClientObject(), client.Merge); err != nil {
+		return res, fmt.Errorf("updating ObjectSet status: %w", err)
+	}
+	return res, nil
+}
+
+// ensures the cache finalizer is set on the given object.
+func (c *GenericObjectSetController) ensureCacheFinalizer(
+	ctx context.Context, objectSet genericObjectSet,
+) error {
+	return controllers.EnsureCommonFinalizer(ctx, objectSet.ClientObject(), c.client, watchFinalizer)
+}
```

Comment:
It seems like sometimes it is called `cacheFinalizer` and sometimes it is called `watchFinalizer`. Are these two different things? Otherwise I think one name might be more clear

Reply:
It was just the watch in the previous codebase and now we have a full cache implementation.
Will make sure cache is used everywhere - too much copy pasta :)